### PR TITLE
Remote mirrors

### PR DESCRIPTION
**Pull request by lukemgriffith**
Adding remote mirrors service, as per the remote mirrors documentation on gitlab.

https://docs.gitlab.com/ce/api/remote_mirrors.html

### CODE REVIEW

File: project_mirror.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+package gitlab
+
+import (
+	"fmt"
+	"time"
+)
+
+// ProjectMirrorService handles communication with the project mirror
+// related methods of the GitLab API.
```
Comment (svanharmelen) [inline project_mirror.go:25]:
```suggestion
// related methods of the GitLab API.
//
// GitLAb API docs:
// https://docs.gitlab.com/ce/api/remote_mirrors.html
```

File: project_mirror.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// ProjectMirror represents a project mirror configuration.
+//
+// GitLAb API docs:
+// https://docs.gitlab.com/ce/api/remote_mirrors.html
+type ProjectMirror struct {
+	Enabled                bool        `json:"enabled"`
+	ID                     int         `json:"id"`
+	LastError              interface{} `json:"last_error"`
```
Comment (svanharmelen) [inline project_mirror.go:37]: Why is this an `interface{}`? This should be a specific type.

File: project_mirror.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	LastSuccessfulUpdateAt time.Time   `json:"last_successful_update_at"`
+	LastUpdateAt           time.Time   `json:"last_update_at"`
+	LastUpdateStartedAt    time.Time   `json:"last_update_started_at"`
```
Comment (svanharmelen) [inline project_mirror.go:40]: Are all these times always returned by the API?

File: project_mirror.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	OnlyProtectedBranches  bool        `json:"only_protected_branches"`
+	KeepDivergentRefs      bool        `json:"keep_divergent_refs"`
+	UpdateStatus           string      `json:"update_status"`
+	URL                    string      `json:"url"`
+}
+
+// ListProjectMirror gets a list of mirrors configured on the project.
+// these are copies of the repositories to external version control repositories
```
Comment (svanharmelen) [inline project_mirror.go:48]:
```suggestion
```

File: project_mirror.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	u := fmt.Sprintf("projects/%s/remote_mirrors", pathEscape(project))
+
+	req, err := s.client.NewRequest("GET", u, nil, options)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var pm []*ProjectMirror
+
```
Comment (svanharmelen) [inline project_mirror.go:66]:
```suggestion
```

File: project_mirror.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// AddProjectMirrorOptions contains the properties requires to create
+// a new project mirror.
+type AddProjectMirrorOptions struct {
+	URL                   string `json:"url"`
+	Enabled               bool   `json:"enabled"`
+	OnlyProtectedBranches bool   `json:"only_protected_branches"`
+	KeepDivergentRefs     bool   `json:"keep_divergent_refs"`
```
Comment (svanharmelen) [inline project_mirror.go:82]: These all need to be pointers and also have `url` struct tags.

File: project_mirror.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return pm, resp, err
+
+}
+
+// AddProjectMirrorOptions contains the properties requires to create
+// a new project mirror.
```
Comment (svanharmelen) [inline project_mirror.go:77]:
```suggestion
// a new project mirror.
//
// GitLab API docs:
// https://docs.gitlab.com/ce/api/remote_mirrors.html#create-a-remote-mirror
```

File: project_mirror.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+// https://docs.gitlab.com/ce/api/remote_mirrors.html#list-a-projects-remote-mirrors
+func (s *ProjectMirrorService) ListProjectMirror(pid interface{}, options ...RequestOptionFunc) ([]*ProjectMirror, *Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+	u := fmt.Sprintf("projects/%s/remote_mirrors", pathEscape(project))
+
+	req, err := s.client.NewRequest("GET", u, nil, options)
+
```
Comment (svanharmelen) [inline project_mirror.go:60]:
```suggestion
```

File: project_mirror.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	resp, err := s.client.Do(req, pm)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return pm, resp, err
+}
+
+// EditProjectMirrorOptions contains the properties requires to edit
+// an existing project mirror.
```
Comment (svanharmelen) [inline project_mirror.go:111]:
```suggestion
// an existing project mirror.
//
// GitLab API docs:
// https://docs.gitlab.com/ce/api/remote_mirrors.html#update-a-remote-mirrors-attributes
```

File: project_mirror.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+type EditProjectMirrorOptions struct {
+	MirrorID              int    `json:"mirror_id"`
+	URL                   string `json:"url"`
+	Enabled               bool   `json:"enabled"`
+	OnlyProtectedBranches bool   `json:"only_protected_branches"`
+	KeepDivergentRefs     bool   `json:"keep_divergent_refs"`
```
Comment (svanharmelen) [inline project_mirror.go:117]: These all need to be pointers and also have url struct tags.

File: project_mirror.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+		return nil, resp, err
+	}
+
+	return pm, resp, err
+}
+
+// EditProjectMirrorOptions contains the properties requires to edit
+// an existing project mirror.
+type EditProjectMirrorOptions struct {
+	MirrorID              int    `json:"mirror_id"`
```
Comment (svanharmelen) [inline project_mirror.go:113]: This one should not be here as it's already in the function signature.

File: project_mirror.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+	}
+	u := fmt.Sprintf("projects/%s/members/%d", pathEscape(project), user)
+
+	req, err := s.client.NewRequest("DELETE", u, nil, options)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.client.Do(req, nil)
+}
```
Comment (svanharmelen) [inline project_mirror.go:162]: This seems to be moved from the project members service, but I don't think we should do that. Or was this meant as a copy/paste/adjust which went bad?

File: project_members.go [hunk 13]
```diff
@@ -188,22 +190,3 @@ func (s *ProjectMembersService) EditProjectMember(pid interface{}, user int, opt
...
-	}
-	u := fmt.Sprintf("projects/%s/members/%d", pathEscape(project), user)
-
-	req, err := s.client.NewRequest("DELETE", u, nil, options)
-	if err != nil {
-		return nil, err
-	}
-
-	return s.client.Do(req, nil)
-}
```
Comment (svanharmelen) [inline project_members.go:209]: I don't think this should be removed, right?

File: project_members.go [hunk 14]
```diff
@@ -14,6 +14,8 @@
 // limitations under the License.
 //
 
+// N.B Documentation does not support remove functionality, alternative is to disable an existing mirror.
```
Comment (svanharmelen) [inline project_members.go:17]: I would not add this comment here. If needed it should be placed in the comment of the related method.

File: project_mirror.go (hunk 3 above)
Comment (lukemgriffith) [reply project_mirror.go]: Documentation doesn't state it, but testing myself looks like these properties are always returned.

**Comment by lukemgriffith** [comment]
Fixed up where requested, there isn't any delete action on the api spec, so I've not included a delete method.

**Comment by lukemgriffith** [comment]
Excellent, thanks.

**Pull request merged**